src/toil/cwl/cwltoil.py

```diff
@@ -1004,6 +1006,22 @@ def run_jobs(
     ) -> None:
         """run_jobs from SingleJobExecutor, but not in a top level runtime context."""
         runtime_context.toplevel = False
+        if isinstance(process, cwltool.command_line_tool.CommandLineTool) and isinstance(process.make_job_runner(runtime_context), SingularityCommandLineJob):
+            # Set defaults for singularity cache environment variables, similar to what we do in wdltoil
+            # Use the same place as the default singularity cache directory
+            singularity_cache = os.path.join(os.path.expanduser("~"), ".singularity")
+            os.environ['SINGULARITY_CACHEDIR'] = os.environ.get("SINGULARITY_CACHEDIR", singularity_cache)
```

Comment:
This code is copied over from the equivalent section from toil-wdl-runner:
https://github.com/DataBiosphere/toil/blob/3ac06815d10f55bfa69306f1f0174cdb5cfffc27/src/toil/wdl/wdltoil.py#L2061-L2068

However, since [Singularity does not want simultaneous pulls no matter what](https://github.com/DataBiosphere/toil/issues/3817), this locking mechanism in both toil-cwl-runner and toil-wdl-runner only works on the Toil side; if there is another singularity pull to the default cache directory from the (same?) user that isn't Toil managed/controlled, then we'd run into the same issue as before. Maybe setting the default cache directory to a toil specific directory like `$HOME/.singularity/toil` would be better?

Cleaning the cache would be another issue as these environment variables do not persist outside of the job.